File: protocols/proven-backup/ffi/zig/src/backup.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// backup.zig -- Zig FFI implementation of proven-backup.
//
// Implements the backup server state machine with:
//   - 64-slot mutex-protected backup job pool
//   - Per-job configuration (type, schedule, compression, encryption)
//   - Job lifecycle: Idle -> Running -> Verifying -> Complete
//   - Failure and cancellation paths
//   - Retention policy per job
//   - Byte counter for progress tracking
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching BackupABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching BackupABI.Types.idr tag assignments)
// =========================================================================

/// Backup types (ABI tags 0-4).
pub const BackupType = enum(u8) {
    full = 0,
    incremental = 1,
    differential = 2,
    snapshot = 3,
    mirror = 4,
};

/// Schedule frequencies (ABI tags 0-4).
pub const ScheduleFreq = enum(u8) {
    hourly = 0,
    daily = 1,
    weekly = 2,
    monthly = 3,
    on_demand = 4,
};

/// Compression algorithms (ABI tags 0-4).
pub const CompressionAlg = enum(u8) {
    none = 0,
    gzip = 1,
    zstd = 2,
    lz4 = 3,
    xz = 4,
};

/// Encryption algorithms (ABI tags 0-2).
pub const EncryptionAlg = enum(u8) {
    no_encryption = 0,
    aes256gcm = 1,
    chacha20poly1305 = 2,
};

/// Backup job lifecycle states (ABI tags 0-5).
pub const BackupState = enum(u8) {
    idle = 0,
    running = 1,
    verifying = 2,
    complete = 3,
    failed = 4,
    cancelled = 5,
};

/// Retention policies (ABI tags 0-4).
pub const RetentionPolicy = enum(u8) {
    keep_all = 0,
    keep_last = 1,
    keep_daily = 2,
    keep_weekly = 3,
    keep_monthly = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent backup jobs.
const MAX_JOBS: usize = 64;

/// A backup job.
const Job = struct {
    /// Current job lifecycle state.
    state: BackupState,
    /// Type of backup.
    backup_type: BackupType,
    /// Schedule frequency.
    schedule: ScheduleFreq,
    /// Compression algorithm.
    compression: CompressionAlg,
    /// Encryption algorithm.
    encryption: EncryptionAlg,
    /// Retention policy.
    retention: RetentionPolicy,
    /// Bytes processed so far.
    bytes_processed: u64,
    /// Whether this job slot is in use.
    active: bool,
};

/// Default (empty) job.
const empty_job: Job = .{
    .state = .idle,
    .backup_type = .full,
    .schedule = .on_demand,
    .compression = .none,
    .encryption = .no_encryption,
    .retention = .keep_all,
    .bytes_processed = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var jobs: [MAX_JOBS]Job = [_]Job{empty_job} ** MAX_JOBS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_JOBS) return null;
    const idx: usize = @intCast(slot);
    if (!jobs[idx].active) return null;
    return idx;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

// -- ABI version --------------------------------------------------------------

/// Returns the ABI version number. Must match Foreign.abiVersion in Idris2.
pub export fn backup_abi_version() callconv(.c) u32 {
    return 1;
}

// -- Lifecycle ----------------------------------------------------------------

/// Create a new backup job. Returns slot index (>=0) or -1 on failure.
/// The job starts in Idle state.
pub export fn backup_create(
    backup_type: u8,
    schedule: u8,
    compression: u8,
    encryption: u8,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (backup_type > 4) return -1;
    if (schedule > 4) return -1;
    if (compression > 4) return -1;
    if (encryption > 2) return -1;

    for (&jobs, 0..) |*j, i| {
        if (!j.active) {
            j.* = empty_job;
            j.backup_type = @enumFromInt(backup_type);
            j.schedule = @enumFromInt(schedule);
            j.compression = @enumFromInt(compression);
            j.encryption = @enumFromInt(encryption);
            j.state = .idle;
            j.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a backup job, releasing its slot.
pub export fn backup_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_JOBS) return;
    jobs[@intCast(slot)] = empty_job;
}

// -- State queries ------------------------------------------------------------

/// Returns the current BackupState tag.
pub export fn backup_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0; // idle fallback
    return @intFromEnum(jobs[idx].state);
}

// -- Job lifecycle transitions ------------------------------------------------

/// Start a backup job. Returns 0 on success, 1 on rejection.
/// Transitions: Idle -> Running.
pub export fn backup_start(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (jobs[idx].state != .idle) return 1;

    jobs[idx].state = .running;
    jobs[idx].bytes_processed = 0;
    return 0;
}

/// Begin verification. Returns 0 on success, 1 on rejection.
/// Transitions: Running -> Verifying.
pub export fn backup_verify(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (jobs[idx].state != .running) return 1;

    jobs[idx].state = .verifying;
    return 0;
}

/// Complete the backup. Returns 0 on success, 1 on rejection.
/// Transitions: Verifying -> Complete.
pub export fn backup_complete(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (jobs[idx].state != .verifying) return 1;

    jobs[idx].state = .complete;
    return 0;
}

/// Fail the backup. Returns 0 on success, 1 on rejection.
/// Transitions: Running/Verifying -> Failed.
pub export fn backup_fail(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = jobs[idx].state;
    if (state != .running and state != .verifying) return 1;

    jobs[idx].state = .failed;
    return 0;
}

/// Cancel the backup. Returns 0 on success, 1 on rejection.
/// Transitions: Running -> Cancelled.
pub export fn backup_cancel(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (jobs[idx].state != .running) return 1;

    jobs[idx].state = .cancelled;
    return 0;
}

// -- Retention policy ---------------------------------------------------------

/// Set retention policy. Returns 0 on success, 1 on rejection.
pub export fn backup_set_retention(slot: c_int, policy: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (policy > 4) return 1;

    jobs[idx].retention = @enumFromInt(policy);
    return 0;
}

/// Returns the current RetentionPolicy tag.
pub export fn backup_retention(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0; // keep_all fallback
    return @intFromEnum(jobs[idx].retention);
}

// -- Progress tracking --------------------------------------------------------

/// Returns bytes processed so far.
pub export fn backup_bytes_processed(slot: c_int) callconv(.c) u64 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return jobs[idx].bytes_processed;
}

// -- Reset --------------------------------------------------------------------

/// Reset a completed/failed/cancelled job to Idle. Returns 0 on success, 1 on rejection.
/// Transitions: Complete/Failed/Cancelled -> Idle.
pub export fn backup_reset(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = jobs[idx].state;
    if (state != .complete and state != .failed and state != .cancelled) return 1;

    jobs[idx].state = .idle;
    jobs[idx].bytes_processed = 0;
    return 0;
}

// -- Stateless transition table -----------------------------------------------

/// Check if a job state transition is valid.
pub export fn backup_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Running
    if (from == 1 and to == 2) return 1; // Running -> Verifying
    if (from == 2 and to == 3) return 1; // Verifying -> Complete
    if (from == 1 and to == 4) return 1; // Running -> Failed
    if (from == 2 and to == 4) return 1; // Verifying -> Failed
    if (from == 1 and to == 5) return 1; // Running -> Cancelled
    if (from == 3 and to == 0) return 1; // Complete -> Idle (reset)
    if (from == 4 and to == 0) return 1; // Failed -> Idle (reset)
    if (from == 5 and to == 0) return 1; // Cancelled -> Idle (reset)
    return 0;
}

File: bindings/rust/src/backup.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//

//! Backup Server types for the proven-servers ABI.
//!
//! Formally verified backup/restore types.
//! Mirrors the Idris2 module `BackupABI.Types`.
//!
//! - `BackupType` -- Backup types.
//! - `ScheduleFreq` -- Backup schedule frequencies.
//! - `CompressionAlg` -- Backup compression algorithms.
//! - `EncryptionAlg` -- Backup encryption algorithms.
//! - `BackupState` -- Backup job states.
//! - `RetentionPolicy` -- Backup retention policies.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// BackupType (tags 0-4)
// ===========================================================================

/// Backup types.
///
/// Matches `BackupType` in `BackupABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum BackupType {
    /// Full (tag 0).
    Full = 0,
    /// Incremental (tag 1).
    Incremental = 1,
    /// Differential (tag 2).
    Differential = 2,
    /// Snapshot (tag 3).
    Snapshot = 3,
    /// Mirror (tag 4).
    Mirror = 4,
}

impl BackupType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Full),
            1 => Some(Self::Incremental),
            2 => Some(Self::Differential),
            3 => Some(Self::Snapshot),
            4 => Some(Self::Mirror),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [BackupType; 5] = [
        Self::Full, Self::Incremental, Self::Differential, Self::Snapshot, Self::Mirror,
    ];
}

impl fmt::Display for BackupType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ScheduleFreq (tags 0-4)
// ===========================================================================

/// Backup schedule frequencies.
///
/// Matches `ScheduleFreq` in `BackupABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ScheduleFreq {
    /// Hourly (tag 0).
    Hourly = 0,
    /// Daily (tag 1).
    Daily = 1,
    /// Weekly (tag 2).
    Weekly = 2,
    /// Monthly (tag 3).
    Monthly = 3,
    /// OnDemand (tag 4).
    OnDemand = 4,
}

impl ScheduleFreq {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Hourly),
            1 => Some(Self::Daily),
            2 => Some(Self::Weekly),
            3 => Some(Self::Monthly),
            4 => Some(Self::OnDemand),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ScheduleFreq; 5] = [
        Self::Hourly, Self::Daily, Self::Weekly, Self::Monthly, Self::OnDemand,
    ];
}

impl fmt::Display for ScheduleFreq {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// CompressionAlg (tags 0-4)
// ===========================================================================

/// Backup compression algorithms.
///
/// Matches `CompressionAlg` in `BackupABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum CompressionAlg {
    /// None (tag 0).
    None = 0,
    /// Gzip (tag 1).
    Gzip = 1,
    /// Zstd (tag 2).
    Zstd = 2,
    /// LZ4 (tag 3).
    Lz4 = 3,
    /// XZ (tag 4).
    Xz = 4,
}

impl CompressionAlg {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::None),
            1 => Some(Self::Gzip),
            2 => Some(Self::Zstd),
            3 => Some(Self::Lz4),
            4 => Some(Self::Xz),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [CompressionAlg; 5] = [
        Self::None, Self::Gzip, Self::Zstd, Self::Lz4, Self::Xz,
    ];
}

impl fmt::Display for CompressionAlg {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// EncryptionAlg (tags 0-2)
// ===========================================================================

/// Backup encryption algorithms.
///
/// Matches `EncryptionAlg` in `BackupABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum EncryptionAlg {
    /// NoEncryption (tag 0).
    NoEncryption = 0,
    /// AES-256-GCM (tag 1).
    Aes256Gcm = 1,
    /// ChaCha20Poly1305 (tag 2).
    ChaCha20Poly1305 = 2,
}

impl EncryptionAlg {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::NoEncryption),
            1 => Some(Self::Aes256Gcm),
            2 => Some(Self::ChaCha20Poly1305),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [EncryptionAlg; 3] = [
        Self::NoEncryption, Self::Aes256Gcm, Self::ChaCha20Poly1305,
    ];
}

impl fmt::Display for EncryptionAlg {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// BackupState (tags 0-5)
// ===========================================================================

/// Backup job states.
///
/// Matches `BackupState` in `BackupABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum BackupState {
    /// Idle (tag 0).
    Idle = 0,
    /// Running (tag 1).
    Running = 1,
    /// Verifying (tag 2).
    Verifying = 2,
    /// Complete (tag 3).
    Complete = 3,
    /// Failed (tag 4).
    Failed = 4,
    /// Cancelled (tag 5).
    Cancelled = 5,
}

impl BackupState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Idle),
            1 => Some(Self::Running),
            2 => Some(Self::Verifying),
            3 => Some(Self::Complete),
            4 => Some(Self::Failed),
            5 => Some(Self::Cancelled),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [BackupState; 6] = [
        Self::Idle, Self::Running, Self::Verifying, Self::Complete, Self::Failed, Self::Cancelled,
    ];
}

impl fmt::Display for BackupState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// RetentionPolicy (tags 0-4)
// ===========================================================================

/// Backup retention policies.
///
/// Matches `RetentionPolicy` in `BackupABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum RetentionPolicy {
    /// KeepAll (tag 0).
    KeepAll = 0,
    /// KeepLast (tag 1).
    KeepLast = 1,
    /// KeepDaily (tag 2).
    KeepDaily = 2,
    /// KeepWeekly (tag 3).
    KeepWeekly = 3,
    /// KeepMonthly (tag 4).
    KeepMonthly = 4,
}

impl RetentionPolicy {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::KeepAll),
            1 => Some(Self::KeepLast),
            2 => Some(Self::KeepDaily),
            3 => Some(Self::KeepWeekly),
            4 => Some(Self::KeepMonthly),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [RetentionPolicy; 5] = [
        Self::KeepAll, Self::KeepLast, Self::KeepDaily, Self::KeepWeekly, Self::KeepMonthly,
    ];
}

impl fmt::Display for RetentionPolicy {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn backup_type_roundtrip() {
        for v in BackupType::ALL {
            let tag = v.to_tag();
            let decoded = BackupType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(BackupType::from_tag(5).is_none());
    }

    #[test]
    fn schedule_freq_roundtrip() {
        for v in ScheduleFreq::ALL {
            let tag = v.to_tag();
            let decoded = ScheduleFreq::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ScheduleFreq::from_tag(5).is_none());
    }

    #[test]
    fn compression_alg_roundtrip() {
        for v in CompressionAlg::ALL {
            let tag = v.to_tag();
            let decoded = CompressionAlg::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(CompressionAlg::from_tag(5).is_none());
    }

    #[test]
    fn encryption_alg_roundtrip() {
        for v in EncryptionAlg::ALL {
            let tag = v.to_tag();
            let decoded = EncryptionAlg::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(EncryptionAlg::from_tag(3).is_none());
    }

    #[test]
    fn backup_state_roundtrip() {
        for v in BackupState::ALL {
            let tag = v.to_tag();
            let decoded = BackupState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(BackupState::from_tag(6).is_none());
    }

    #[test]
    fn retention_policy_roundtrip() {
        for v in RetentionPolicy::ALL {
            let tag = v.to_tag();
            let decoded = RetentionPolicy::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(RetentionPolicy::from_tag(5).is_none());
    }

}
